engine/src/main/java/com/arcadedb/serializer/BinarySerializer.java

```diff
@@ -563,10 +566,26 @@ else if (value instanceof LocalDate)
     default:
       LogManager.instance().log(this, Level.INFO, "Error on serializing value '" + value + "', type not supported");
     }
+
+    if (dataEncryption != null) {
+      switch (type) {
+        case BinaryTypes.TYPE_NULL:
+        case BinaryTypes.TYPE_COMPRESSED_RID:
+        case BinaryTypes.TYPE_RID:
+          break;
+        default:
+          serialized.putBytes(dataEncryption.encrypt(content.toByteArray()));
```

Comment:
I see you're encrypting every single value. Why don't encrypting the whole Binary at the end, once all the values are serialized? This would be much faster and you don't need to differentiate special cases like RIDs. WDYT?

Reply:
Thanks, I'll have a look

Reply:
I was looking inside the code to find best place to do it, as `Binary` is referred in plenty of places, and it seemed that interactions with `BaseRecord#buffer` is what I am looking for. That however also is fairly nested inside many classes meaning it is difficult to intercept right moment to perform write/read as it gets passed around and often `buffer` is accessed directly to perform partial read.

In the end, I chosen implementing encryption at `BinarySerializer` in `serializeProperties()`, `deserializeProperties()`, `deserializeProperty` and `hasProperty()`. Idea is to write just all properties as encrypted at once and vice-versa. The issue I ran into is that `serializeProperties()` writes to two types: `header` and `content`. Header stores information about `propertiesCount` and about each `propertyKey` with its `bufferPosition` for the value. This allows reading just the value from whole buffer, and not all properties to find one. However this isn't possible with the goal above. I have to either re-factor the code to read all properties from buffer to decode, regardless of looking for specific one and then filter by key, or leave existing behaviour.

I could also re-factor structure of `BaseRecord#binary` to distinct between meta content (like ID) and data (encrypted) but that could be even more re-factoring, and it wouldn't resolve cherry picking properties anyway.

I understand that current implementation with encryption of each value adds performance cost, I didn't measure how this would affect our product yet, but it is mandatory requirement for us anyway.

What do you think?

Reply:
Sorry for the delay on this review. It makes sense, it's better to avoid encrypting the metadata for fast accessing to the properties you're requesting. Also, this makes easier to encrypt single property by adding some configuration (in the Schema -> Type -> Property)